packages/ozone/src/verification/issuer.ts

```diff
@@ -0,0 +1,119 @@
+import { Verification } from '../db/schema/verification'
+
+export type VerificationInput = {
+  displayName: string
+  handle: string
+  subject: string
+  createdAt?: string
+}
+
+export type VerificationIssuerCreator = (
+  verifierConfig: VerifierConfig,
+) => VerificationIssuer
+
+export class VerificationIssuer {
+  constructor(private verifierConfig: VerifierConfig) {}
+
+  static creator() {
+    return (verifierConfig: VerifierConfig) =>
+      new VerificationIssuer(verifierConfig)
+  }
+
+  // @TODO: Probably shouldn't login for every req?
+  async getAgent() {
+    const agent = new AtpAgent({ service: this.verifierConfig.url })
+    await agent.login({
+      identifier: this.verifierConfig.did,
+      password: this.verifierConfig.password,
+    })
+
+    return agent
+  }
+
+  async verify(verifications: VerificationInput[]) {
+    const grantedVerifications: Selectable<Verification>[] = []
+    const failedVerifications: {
+      $type: 'tools.ozone.verification.grant#grantError'
+      subject: string
+      error: string
+    }[] = []
+    const now = new Date().toISOString()
+    const agent = await this.getAgent()
+    await Promise.allSettled(
+      verifications.map(async ({ displayName, handle, subject, createdAt }) => {
+        try {
+          const verificationRecord = {
+            createdAt: createdAt || now,
+            issuer: this.verifierConfig.did,
+            displayName,
+            handle,
+            subject,
+          }
+          const {
+            data: { uri },
+          } = await agent.com.atproto.repo.createRecord({
+            repo: this.verifierConfig.did,
+            record: verificationRecord,
+            collection: 'app.bsky.graph.verification',
+          })
+          grantedVerifications.push({
+            ...verificationRecord,
+            uri,
+            revokedAt: null,
+            updatedAt: now,
+            revokedBy: null,
+            revokeReason: null,
+          })
+        } catch (err) {
+          failedVerifications.push({
+            $type: 'tools.ozone.verification.grant#grantError',
+            error: (err as Error).message,
+            subject,
+          })
+          return
+        }
+      }),
+    )
+
+    return { grantedVerifications, failedVerifications }
+  }
+
+  async revoke({ uris }: { uris: string[] }) {
+    const revokedVerifications: string[] = []
+    const failedRevocations: string[] = []
+
+    await Promise.allSettled(
+      uris.map(async (uri) => {
+        try {
+          const atUri = new AtUri(uri)
+
+          if (atUri.collection !== 'app.bsky.graph.verification') {
+            throw new Error(`Only verification records can be revoked`)
+          }
+
+          if (atUri.host !== this.verifierConfig.did) {
+            throw new Error(
+              `Cannot revoke verification record ${uri} not issued by ${this.verifierConfig.did}`,
+            )
+          }
+
+          const agent = await this.getAgent()
```

Comment:
Given that `await this.getAgent()` currently logs in for every call, then we should at least share a single agent for the whole batch.  I.e. not create 100 sessions for a single batch of 100.

Reply:
yep yep. i had that outside of the loop for the record creator but missed this one.